spec/lib/zendesk_apps_tools/api_connection_spec.rb

```diff
@@ -0,0 +1,115 @@
+        it 'replaces %s with subdomain in template' do
+          user_input_subdomain = 'my-subdomain'
+          expect(default_url_template % user_input_subdomain).to eq("https://my-subdomain.zendesk.com/")
+        end
+      end
+    end
+
+    describe 'SUBDOMAIN_VALIDATION_PATTERN' do
+      context 'valid_subdomain? (a private method)' do
+        define_method(:valid_subdomain?) { |subdomain| !!subdomain_validation_pattern.match(subdomain) }
+
+        it 'returns false if subdomain is NOT in valid format' do
+          expect(valid_subdomain?('sub.domain')).to eq(false)
+          expect(valid_subdomain?('sub!domain')).to eq(false)
+          expect(valid_subdomain?('sub~domain')).to eq(false)
+          expect(valid_subdomain?('sub_domain')).to eq(false)
+        end
+
+        it 'returns true if subdomain is in valid format' do
+          expect(valid_subdomain?('subDomain')).to eq(true)
+          expect(valid_subdomain?('SUBDOMAIN')).to eq(true)
+          expect(valid_subdomain?('subdomain')).to eq(true)
+          expect(valid_subdomain?('sub-domain')).to eq(true)
+        end
+      end
+    end
+
+    describe 'ZENDESK_URL_VALIDATION_PATTERN' do
+      context 'valid_full_url? (a private method)' do
+        define_method(:valid_full_url?) { |subdomain|  !!url_validation_pattern.match(subdomain) }
+
+        context 'with regular zendesk urls' do
+          it 'returns false when subdomain does not match full url pattern' do
+            expect(valid_full_url?('www.subdomain.com')).to eq(false)
+            expect(valid_full_url?('subdomain.com')).to eq(false)
+          end
+
+          it 'returns true when subdomain does match full url pattern' do
+            expect(valid_full_url?('http://z3n-subdomain.zendesk.com')).to eq(true)
+            expect(valid_full_url?('https://subdomain.zendesk.com')).to eq(true)
+            expect(valid_full_url?('https://my-subdomain.zendesk-staging.com')).to eq(true)
+          end
+        end
+
+        context 'with host map urls' do
+          it 'returns true when subdomain of customized urls matches full url pattern' do
+            expect(valid_full_url?('https://subdomain.com')).to eq(true)
+            expect(valid_full_url?('https://www.subdomain.com')).to eq(true)
+            expect(valid_full_url?('https://subdomain.au')).to eq(true)
+          end
+        end
+      end
+    end
+  end
+
+  describe '#prepare_api_auth' do
+    let(:url_error_message) { ZendeskAppsTools::APIConnection::URL_ERROR_MSG }
+    let(:email_error_message) { ZendeskAppsTools::APIConnection::EMAIL_ERROR_MSG }
+    let(:subject_class) do
+      Class.new do
+        include ZendeskAppsTools::APIConnection
+        attr_reader :cache, :subdomain, :username
+
+        def initialize(subdomain = nil, username = nil)
+          zat_cache = {
+            'subdomain' => @subdomain = subdomain,
+            'username' => @username = username
+          }
+          @cache = zat_cache
```

Comment:
@token-cjg Another thing to add to our session today,

The reason why I created `cache` as an attribute accessor is because of line 18 and 21 in `api_connection.rb`, where it goes `cache.fetch` `subdomain`/`username`. 

`@cache`here would replicate cache in ZAT

Reply:
<img width="393" alt="screen shot 2018-12-04 at 4 17 31 pm" src="https://user-images.githubusercontent.com/17760485/49420562-2af16c00-f7e0-11e8-916a-145234c52e6d.png">

Reply:
Oh, interesting!